[otlp] UseOtlpExporter cross-cutting extension

CodeBlanch opened the pull request:
Fixes #4940
Relates to #5325

## Changes

* Adds a cross-cutting `UseOtlpExporter` extension targeting `IOpenTelemetryBuilder`. This extension behaves as follows:
  * Adds an `OtlpExporter` for logging, metrics, & tracing at the end of the processor pipeline for logs & tracing.
  * Automatically turns on Logging, Metrics, & Tracing (performs "With" calls).
    * Note: Calls to `AddSource` \ `AddMeter` are still required to listen when using tracing & metrics.
  * More than one call to `UseOtlpExporter` will result in `NotSupportedException` being thrown.
  * Causes the signal-specific `AddOtlpExporter` extensions to throw `NotSupportedException` once called.
* Adds support for the signal-specific environment variables outlined in the spec when using `UseOtlpExporter`.

## Scenarios

### Scenario 1:

Add OtlpExporter for all signals (at the end of pipelines) using defaults + envvars to do configuration:

```csharp
appBuilder.Services.AddOpenTelemetry()
    .ConfigureResource(r => r.AddService("MyService"))
    .UseOtlpExporter();
```

### Scenario 2:

Add OtlpExporter for all signals (at the end of pipelines) and specify common settings:

```csharp
appBuilder.Services.AddOpenTelemetry()
    .ConfigureResource(r => r.AddService("MyService"))
    .UseOtlpExporter(OtlpExportProtocol.HttpProtobuf, new Uri("http://my_otlp_endpoint/"));
```

### Scenario 3:

Use the cross-cutting UseOtlpExporter method with some pipeline customizations.

This call order...

```csharp
appBuilder.Services.AddOpenTelemetry()
    .ConfigureResource(r => r.AddService("MyService"))
    .UseOtlpExporter()
    .WithTracing(tracing => tracing
        .AddAspNetCoreInstrumentation()
        .AddProcessor(new MyRedactionProcessor)
        .AddConsoleExporter());
```

...will result in a tracing processor pipeline like this:

`MyRedactionProcessor -> SimpleExportProcessor[ConsoleExporter] -> BatchExportProcessor[OtlpExporter]`

Because `UseOtlpExporter` adds its processor/exporter to the end of the pipeline.

There will be some things which won't work as a result. Such as [AutoFlushActivityProcessor](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Extensions#autoflushactivityprocessor), which needs to be at the end.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

samsp-msft commented:
Tagging: @davidfowl, @noahfalk

noahfalk commented:
For scenario 4 I agree with Sam that the API seems more complex than AddOtlpExporter() per-signal. What if we just said global AddOtlpExporter() is only for cases 1,2, and 3 and anyone who wants (4) will need to call the signal specific AddOtlpExporter()? My suggestion is close to Sam's but without attempting to use global AddOtlpExporter() as a default config that can be overriden per-signal. I'd suggest that if users did this:
```C#
appBuilder.Services.AddOpenTelemetry()
    .ConfigureResource(r => r.AddService("MyService"))
    .AddOtlpExporter(...)
    .WithTracing()
    .WithMetrics().AddOtlpExporter(...)
    .WithLogging();
```
We just give them two exporters for the metrics pipeline, one with the global config and one with the local config. If both of the exporters are configured to send data to the same place perhaps we can give a warning or error unless there is some scenario where its reasonable you'd want exporters configured that way?


For scenario 5 IMO it would be fine to say that users who don't want the OTLP exporter at the end of the pipeline shouldn't use the global AddOtlpExporter() method. Instead they would need to configure each pipeline individually in the order they want it.

For scenario 6 I think its fine not to support it. If in the future you did want to support something like that perhaps this is the one that looks like Sam's suggested global defaults that are overriden per signal. Something like:
```C#
appBuilder.Services.AddOpenTelemetry()
    .ConfigureResource(r => r.AddService("MyService"))
    .ConfigureOtlpExporterDefaults(...)
    .WithTracing().AddOtlpExporter(...)
    .WithMetrics().AddOtlpExporter(...)
    .WithLogging().AddOtlpExporter(...);
```

samsp-msft commented:
@CodeBlanch - I think this PR has gone way beyond what the .NET team was expecting, which is both a blessing and a curse. Can we simplify it down to just supporting scenarios 1, 2 & 3?

For scenario 5 & 6, I don't think we were expecting the instancing to be changing as part of this request. I was expecting the `AddOTLPExporter()` off `AddOpenTelemetry()` to really be a helper method, that would implicitly call `AddOTLPExporter` on each source. The primary goal is to reduce the number of lines of code - especially in scenarios where the configuration is being supplied by OTel environment variables.

I love the configuration aspect of scenario 3, but that could come as a later feature if required. We want this to land for Aspire GA and we are running out of runway.

CodeBlanch commented:
@samsp-msft 

I'm open to anything.

Originally my thinking was just to add `IOpenTelemetryBuilder.AddOtlpExporter()` and let _all_ options comes from env vars. Which is really want Aspire is after as you mentioned. But @reyang raised a good point. That puts us into this design...

* If you need A call the cross-cutting AddOtlpExporter method.
* If you need X, Y, or Z call the signal-specific AddOtlpExporter methods.

Today the story is simple:

* Whatever you need call the signal-specific AddOtlpExporter methods.

Do you think the benefit in folding a few lines of code justifies mucking that story up? Especially given the ordering concerns.

For context what I _tried_ to do with this was make the story:

* Whatever you need call the cross-cutting AddOtlpExporter method.

More or less set us up to obsolete the old stuff because there are a myriad of issues with it.

<details>
<summary>PS: If is actually a bit nicer than that.</summary>

Today the story is actually...

* Whatever you need call the signal-specific AddOtlpExporter methods.
* Signal-specific envvars are not supported.

This PR is actually...

* Whatever you need call the cross-cutting AddOtlpExporter method and it will respect signal-specific envvars.

Why are signal-specific envvars not respected today? Tracing, metrics, & logging all use `OtlpExporterOptions` class and there isn't a way to know when retrieving options which instance is being configured.
</details>

samsp-msft commented:
> @samsp-msft
> 
> I'm open to anything.
> 
> Originally my thinking was just to add `IOpenTelemetryBuilder.AddOtlpExporter()` and let _all_ options comes from env vars. Which is really want Aspire is after as you mentioned. But @reyang raised a good point. That puts us into this design...
> 
> * If you need A call the cross-cutting AddOtlpExporter method.
> * If you need X, Y, or Z call the signal-specific AddOtlpExporter methods.
> 
> Today the story is simple:
> 
> * Whatever you need call the signal-specific AddOtlpExporter methods.
> 
> Do you think the benefit in folding a few lines of code justifies mucking that story up? Especially given the ordering concerns.
> 
> For context what I _tried_ to do with this was make the story:
> 
> * Whatever you need call the cross-cutting AddOtlpExporter method.
> 
> More or less set us up to obsolete the old stuff because there are a myriad of issues with it.
> 
> PS: If is actually a bit nicer than that.

Unfortunately I think the clock is going to be the biggest limitation here as we want a solution that works nicely for Aspire. That technically only requires Scenario 1, but 2 & 3 are really nice to have. The deadline is March 22nd, so what can we deliver for that timeline, and deliver the other scenarios in later releases.

noahfalk commented:
My two cents :)

> Do you think the benefit in folding a few lines of code justifies mucking that story up?

The APIs today may have a simple premise, but it doesn't feel like that premise aligns with how customers intuitively conceptualize it. I expect many users initial concept is "I configure what telemetry to collect and where it should be sent" rather than "I configure three different telemetry pipelines from building blocks of sources, processors, and exporters". I think of this as "Make the code feel intuitive for users" rather than "Help users write fewer lines of code".

My impression is that if we add the new API targeting at least scenarios 1,2 and 3 it makes getting started easier for most people and it adds a step in the learning curve when transitioning from the trivial scenarios to the customized pipeline cases. This feels like a net win to me, especially because I expect most developers will never need to handle those more complex cases.

In order to ease the learning curve for the smaller number of folks who do need to cross over to the custom pipeline case I think we could be very explicit in the documentation and comments that pop up in IDE. Something like "This method is a shortcut that is equivalent to adding the OTLP exporter to the metrics, logging, and tracing pipelines individually. No signal specific customization APIs are available when using this shortcut. If you need those customizations, use AddOtlpExporter() on the TracerProviderBuilder, MeterProviderBuilder, and LoggingProviderBuilder instead of using this method."

> More or less set us up to obsolete the old stuff because there are a myriad of issues with it.

I think the general principle of 'obsolete the old stuff so we can fix serious design problems' is fine. If you still want to support the advanced scenarios of per-signal pipelines where ordering matters and the OTLP exporter needs to not be at the end, that feels a rough fit for this global AddOtlpExporter() method to handle. I think if you wanted to fully obsolete and replace then probably some per-signal OTLP method is needed for those particularly complex scenarios. I'm also guessing that represents one of the most complex cases that is rarely used so it doesn't need to be treated with the same priority.

> The deadline is March 22nd, so what can we deliver for that timeline, and deliver the other scenarios in later releases.

As a quick check, is an OTel change prior to March 22nd potentially in the cards? I don't recall your ship schedule to know if there is a new official build of OTel planned to go out before that date?

CodeBlanch commented:
I just pushed updates after bouncing ideas off a bunch of different folks:

* Renamed the method `UseOtlpExporter` (from `AddOtlpExporter`). This is because it behaves differently from the traditional "Add" methods. Primarily it adds its processor to the end of the pipeline(s). Since it is behaving differently, I also changed it to automatically turn on the signals (no "With" calls are necessary). There is no way to change that "add-to-end" behavior other than go and call the extensions we have today.
* Removed most of the scenarios. The pressing need for Aspire is just envvars. The other stuff we can work on as follow-ups.
* In order to simplify the story, you can't call "UseOtlpExporter" more than once and you can't mix it with signal-specific "AddOtlpExporter" calls. Those will result in `NotSupportedException`s. We can always improve support in the future and loosen those restrictions based on needs that crop up.

reyang commented:
@CodeBlanch can you please add a scenario to clarify the expected behavior when both `UseOtlpExporter` and `With...` are used?

For example:

```csharp
appBuilder.Services.AddOpenTelemetry()
    .ConfigureResource(r => r.AddService("MyService"))
    .UseOtlpExporter()
    .WithTracing(tracing => tracing
        .AddAspNetCoreInstrumentation()
        .AddProcessor(new MyRedactionProcessor)
        .AddConsoleExporter());
```

Specifically, I want to make sure folks understand the ordering challenge, and some scenarios might not work as most users would imagine. For example: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Extensions

CodeBlanch commented:
@reyang Added to PR description

reyang commented:
Now the proposal looks good to me.

reyang commented on src/OpenTelemetry/BaseProcessor.cs:
```diff
@@ -41,7 +43,7 @@ public BaseProcessor()
     internal
 #endif
-        int Weight { get; set; }
+        ProcessorPipelineWeight PipelineWeight { get; set; } = ProcessorPipelineWeight.PipelineMiddle;
```
I think I understand the intention of introducing this concept. I don't feel I would agree/approve it (I have trouble with this direction), and considering time constraints, consider removing it from this PR.

CodeBlanch replied:
There is no way to force something to the back of the pipeline currently. If I remove this, this PR and all of its scenarios need to be reworked. Here are some options:

1) We can make `PipelineWeight` purely internal in SDK and always expose SDK internals to OtlpExporter. No issue with this, just that components (like `AutoFlushActivityProcessor`) can never be made to work together without a way to participate in the ordering. The reason I felt exposing it experimentally was a good idea is we can see if it works and then bring learnings back to the spec. But we could always tackle that whole effort separately.

2) We can come up with some other way to do it. Likely will also need an API though so we could end up back in the same spot.

3) Rework this PR and its scenarios so it doesn't try to add itself to end of the pipeline.

Quickest option would probably be 1.

reyang replied:
I'm fine as long as there is no public API (even experimental one) for "weight".

CodeBlanch replied:
Updated: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5400/commits/c2db1009e350045b1765456c60aee948c3d572d9

vishweshbankwar commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OpenTelemetryBuilderOtlpExporterExtensions.cs:
```diff
@@ -0,0 +1,213 @@
+    /// <param name="baseEndpoint">The base endpoint to use. A signal-specific
+    /// path will be appended to the base endpoint for each signal
+    /// automatically if the protocol is set to <see cref="OtlpExportProtocol.HttpProtobuf"/>.</param>
+    public static IOpenTelemetryBuilder UseOtlpExporter(
```
This looks like grpc protocol specific method to me. Do we need this as a separate method?

CodeBlanch replied:
I'm not following. What do you mean by grpc-specific? At the moment there are 3 overloads exposed:

```csharp
// Everything is default or comes from IConfiguration/envvars
appBuilder.Services.AddOpenTelemetry()
    .UseOtlpExporter();

// Specify the endpoint. Everything else is default or comes from IConfiguration/envvars
appBuilder.Services.AddOpenTelemetry()
    .UseOtlpExporter(new Uri("http://my_otlp_endpoint/"));

// Specify the protocol + endpoint. Everything else is default or comes from IConfiguration/envvars
appBuilder.Services.AddOpenTelemetry()
    .UseOtlpExporter(OtlpExportProtocol.HttpProtobuf, new Uri("http://my_otlp_endpoint/"));
```

vishweshbankwar replied:
for this particular case

```
appBuilder.Services.AddOpenTelemetry()
    .UseOtlpExporter(new Uri("http://my_otlp_endpoint/"));
```

Are expecting many users specifying the endpoint here and changing the protocol via environment variable?

CodeBlanch replied:
@vishweshbankwar 

I chatted with @alanwest a bit. We decided to tweak the overloads. They are now:

```
UseOtlpExporter();
UseOtlpExporter(OtlpExportProtocol protocol);
UseOtlpExporter(OtlpExportProtocol protocol, Uri baseEndpoint);
```

How you feel about that? TBH we weren't sure exactly what concern you were raising but these seemed more useful in the end 😄

CodeBlanch replied:
Chatted offline with @vishweshbankwar. We decided to drop the second overload. What we have now is:

```
UseOtlpExporter();
UseOtlpExporter(OtlpExportProtocol protocol, Uri baseEndpoint);
```

utpilla commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OpenTelemetryBuilderOtlpExporterExtensions.cs:
```diff
@@ -0,0 +1,213 @@
+            otlpBuilder.ConfigureDefaultExporterOptions(o =>
+            {
+                o.Protocol = protocol;
+                if (baseEndpoint != null)
```
Remove this check. You already have a `Guard.ThrowIfNull` for `baseEndpoint`.

vishweshbankwar commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs:
```diff
@@ -0,0 +1,148 @@
+                    }
+                },
+                "MetricsOptions": {
+                    "Endpoint": "http://metrics_endpoint/metrics/"
```
nit: Show MetricReaderOptions also.

utpilla commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OpenTelemetryBuilderOtlpExporterExtensions.cs:
```diff
@@ -0,0 +1,210 @@
+            sp.GetService<IOptionsMonitor<MetricReaderOptions>>()?.Get(name),
+            sp.GetService<IOptionsMonitor<ActivityExportProcessorOptions>>()?.Get(name)));
+
+        name ??= Options.DefaultName;
```
For my understanding, why do we not need the `??=` assignment for the `sp.GetService<T>()?.Get(name)` calls above?

CodeBlanch replied:
That is a mistake! I moved it up: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5400/commits/6f2c8e012d119b7988e3ffe8b43924ce1c713b39

The options code seems smart enough to treat `null` name as `Options.DefaultName` (ex: https://github.com/dotnet/runtime/blob/cbca5083d3e69f2bd25e397f8894d94d7763a13a/src/libraries/Microsoft.Extensions.Options/src/OptionsCache.cs#L35) so we might actually be able to remove it completely. But I'm going to leave in there for now I think better safe than sorry. We do that a lot of spots so maybe a cleanup PR to remove it everywhere at some point, need to verify all the versions of Options API do it correct which is kind of heavy lifting 🤣

utpilla commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OpenTelemetryBuilderOtlpExporterExtensions.cs:
```diff
@@ -0,0 +1,210 @@
+                var processor = OtlpLogExporterHelperExtensions.BuildOtlpLogExporter(
+                    sp,
+                    builderOptions.LoggingOptionsInstance.ApplyDefaults(builderOptions.DefaultOptionsInstance),
+                    builderOptions.LogRecordExportProcessorOptions ?? throw new NotSupportedException(),
```
Could you add a comment here describing what it means to have `builderOptions.LogRecordExportProcessorOptions` as `null` here or why we throw a `NotSupportedException`?

CodeBlanch replied:
I added some notes: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5400/commits/b75f139910eef47fbf56644a0756045713bee107

It is really a left over thing from an earlier iteration where you could select the signals you wanted. But I think it is good to leave just in case we ever restore that?

utpilla commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OpenTelemetryBuilderOtlpExporterExtensions.cs:
```diff
@@ -0,0 +1,216 @@
+        services.ConfigureOpenTelemetryLoggerProvider(
+            (sp, logging) =>
+            {
+                var builderOptions = GetBuilderOptionsAndValidateRegistrations(sp, name);
```
Could we do this once outside of the individual signal configuration? It looks like they all check the same thing and retrieve the same options object?

CodeBlanch replied:
Not really. We need the `IServiceProvider` and there isn't currently a way to get it outside of these provider configuration extensions. Could possibly add something, but probably needs to be its own effort and then we can clean this up.

utpilla commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md:
```diff
@@ -39,6 +39,29 @@
+  The new extension has the following behaviors:
+
+  * Calling `UseOtlpExporter` will automatically enable logging, tracing, and
+    metrics. Additional calls to `WithLogging`, `WithMetrics`, and `WithTracing`
```
> Additional calls to `WithLogging`, `WithMetrics`, and `WithTracing` are NOT required.

Do we configure the TracerProvider and MeterProvider to listen to any `ActivitySource` or `Meter` for the export to export something (when there are no `WithTracing` and `WithMetrics` calls)?

CodeBlanch replied:
Not currently. So there will likely need to be calls to either `WithTracing` / `WithMetrics` or `ConfigureOpenTelemetryTracerProvider` / `ConfigureOpenTelemetryMeterProvider` to turn _something_ on. For metrics users could use the config stuff @noahfalk added in .NET 8 to eliminate those calls.

I tweaked the CHANGELOG a bit: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5400/commits/0faac1a49f79c5b906d5b8d854bc651ef7db9ede

LMK what you think about that.

utpilla replied:
> Automatically turns on Logging, Metrics, & Tracing

Could you update the PR description as well to mention this?

CodeBlanch replied:
Done

utpilla commented on src/OpenTelemetry.Exporter.OpenTelemetryProtocol/IOtlpExporterOptions.cs:
```diff
@@ -0,0 +1,108 @@
+/// Describes the OpenTelemetry Protocol (OTLP) exporter options shared by all
+/// signals.
+/// </summary>
+internal interface IOtlpExporterOptions
```
Is this interface only being added for DI config purposes? If yes, could you add a comment here similar to the one for `UseOtlpExporterRegistration` [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5400/files#diff-56333e99a46dc269315055dd59f70eeb552bf53512014dda2aa09bb864c02921R8-R10)?

CodeBlanch replied:
Not for DI purposes.

We have this thing: https://github.com/open-telemetry/opentelemetry-dotnet/issues/4527

There are some trace settings on `OtlpExporterOptions`. Generally we (me and @alanwest, maybe others) feel this was a mistake. For metrics & logs we went in a different direction. Those two properties on `OtlpExporterOptions` we want to obsolete. #5423 is kind of the groundwork for that.

This PR is adding `OtlpExporterBuilderOptions` which could have looked like this:

```csharp
internal sealed class OtlpExporterBuilderOptions
{
    public OtlpExporterOptions DefaultOptions { get; }
    public OtlpExporterOptions LoggingOptions { get; }
    public OtlpExporterOptions MetricsOptions { get; }
    public OtlpExporterOptions TracingOptions { get; }
}
```

Problem there is it is using `OtlpExporterOptions` which has the trace settings. What I didn't want to do was carry forward that mistake into this new API. What I did instead was:

```csharp
internal sealed class OtlpExporterBuilderOptions
{
    public IOtlpExporterOptions DefaultOptions { get; }
    public IOtlpExporterOptions LoggingOptions { get; }
    public IOtlpExporterOptions MetricsOptions { get; }
    public IOtlpExporterOptions TracingOptions { get; }
}
```

What `IOtlpExporterOptions` does is expose just the stuff that is _really_ supported by all the signals.

Now this is all internal at the moment. I could remove `IOtlpExporterOptions` from this PR. But I would like to keep it so the contract is there. Hoping to expose some additional overloads to `UseOtlpExporter` to add additional scenarios in the future which will require this stuff to be made public (could be experimental).

utpilla replied:
Thanks for the explanation!

utpilla commented on test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/UseOtlpExporterExtensionTests.cs:
```diff
@@ -0,0 +1,344 @@
+
+        using var sp = services.BuildServiceProvider();
+
+        Assert.Throws<NotSupportedException>(() => sp.GetRequiredService<LoggerProvider>());
```
Does this test that the user would actually get an exception when trying to run the application? It's probably some `ServiceProvider` related deferred execution thing but I was hoping for the code to throw an exception while executing the second   call to `UseOtlpExporter`. Maybe inspect the `ServiceCollection` itself before adding `UseOtlpExporterRegistration`?

```csharp
if (services.Any((service) => service.ServiceType == typeof(UseOtlpExporterRegistration)))
{
    throw new NotSupportedException("Multiple calls to UseOtlpExporter on the same IServiceCollection are not supported.");
}

services.AddSingleton<UseOtlpExporterRegistration>();
```

CodeBlanch replied:
Slight problem with doing it this way.

It will work for this case:

```csharp
        services.AddOpenTelemetry()
           .UseOtlpExporter()
           .UseOtlpExporter();
```

But it won't work for this case:

```csharp
        services.AddOpenTelemetry()
           .WithTracing(builder => builder.AddOtlpExporter()) // `UseOtlpExporterRegistration` won't exist in service collection until the next line runs
           .UseOtlpExporter();
```

The question is, will the current mechanism crash the app?

So what will happen is a `NotSupportedException` is thrown when creating either `LoggerProvider`, `MeterProvider`, or `TracerProvider`.

In the case of hosting, that will happen in the `IHostedService`.

I just dropped two `UseOtlpExporter` calls into the `Examples.AspNetCore` app to verify. It does throw:

<details>
<summary>Stack trace</summary>

```
Unhandled exception. System.NotSupportedException: Multiple calls to UseOtlpExporter on the same IServiceCollection are not supported.
   at System.OpenTelemetryBuilderServiceProviderExtensions.EnsureSingleUseOtlpExporterRegistration(IServiceProvider serviceProvider) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\Builder\OpenTelemetryBuilderServiceProviderExtensions.cs:line 16
   at OpenTelemetry.Exporter.OtlpExporterBuilder.<RegisterOtlpExporterServices>g__GetBuilderOptionsAndValidateRegistrations|14_6(IServiceProvider sp, String name) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\Builder\OtlpExporterBuilder.cs:line 248
   at OpenTelemetry.Exporter.OtlpExporterBuilder.<>c__DisplayClass14_0.<RegisterOtlpExporterServices>b__3(IServiceProvider sp, LoggerProviderBuilder logging) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\Builder\OtlpExporterBuilder.cs:line 198
   at OpenTelemetry.Logs.OpenTelemetryDependencyInjectionLoggingServiceCollectionExtensions.ConfigureLoggerProviderBuilderCallbackWrapper.ConfigureBuilder(IServiceProvider serviceProvider, LoggerProviderBuilder loggerProviderBuilder) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Api.ProviderBuilderExtensions\Logs\OpenTelemetryDependencyInjectionLoggingServiceCollectionExtensions.cs:line 175
   at OpenTelemetry.Logs.LoggerProviderSdk..ctor(IServiceProvider serviceProvider, Boolean ownsServiceProvider) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry\Logs\LoggerProviderSdk.cs:line 50
   at OpenTelemetry.Logs.LoggerProviderBuilderBase.<>c.<.ctor>b__3_0(IServiceProvider sp) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry\Logs\Builder\LoggerProviderBuilderBase.cs:line 42
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.Logging.OpenTelemetryLoggingExtensions.<>c.<AddOpenTelemetryInternal>b__5_1(IServiceProvider sp) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry\Logs\ILogger\OpenTelemetryLoggingExtensions.cs:line 246
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.Hosting.HostBuilder.<>c__DisplayClass35_0.<PopulateServiceCollection>b__2(IServiceProvider _)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.Hosting.HostBuilder.ResolveHost(IServiceProvider serviceProvider, DiagnosticListener diagnosticListener)
   at Microsoft.Extensions.Hosting.HostApplicationBuilder.Build()
   at Microsoft.AspNetCore.Builder.WebApplicationBuilder.Build()
   at Program.<Main>$(String[] args) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\examples\AspNetCore\Program.cs:line 164
```
</details>

What if the user isn't using hosting? Well at the moment there isn't a way to get an `IOpenTelemetryBuilder` to use for a call to `UseOtlpExpoter` without hosting 🤣 

#5325 is adding a way to do cross-cutting things without using hosting. If we move forward with that, the current mechanism will crash earlier for that API...

```csharp
     using var openTelemetry = OpenTelemetrySdk.Create(Configure); // NotSupportedException thrown here

     static void Configure(IOpenTelemetryBuilder builder)
     {
            builder
               .ConfigureResource(r => r.AddService("MyService"))
               .UseOtlpExporter();
     }
```

CodeBlanch commented on test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpExporterOptionsTests.cs:
```diff
@@ -259,6 +259,46 @@ public void OtlpExporterOptions_HttpClientFactoryThrowsWhenSetToNull()
+        defaultOptions.TimeoutMilliseconds = 18;
+        defaultOptions.HttpClientFactory = () => null!;
+
+        var signalOptionsWithoutData = new OtlpExporterOptions(OtlpExporterOptionsConfigurationType.Traces);
```
We could, but there's really nothing signal-specific about this test. It is meant to cover the `ApplyDefaults` method. I tweaked it a bit to remove the signals that were in there: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5400/commits/956f537e0f755c01b2e5bb393dbe8fda616bbe27

LMK if that helps with the clarity.

utpilla commented on test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/UseOtlpExporterExtensionTests.cs:
```diff
@@ -0,0 +1,345 @@
+    }
+
+    [Fact]
+    public void UseOtlpExporterSetEndpointAndProtocolTest()
```
Mark this test and `UseOtlpExporterSetProtocolTest` as `Theory` to test both the protocols.

CodeBlanch replied:
I went ahead and made this change but testing `OtlpExportProtocol.Grpc` is a bit tricky because that is the default value. The test really can't know for certain if the protocol was applied or just the default was used. With `OtlpExportProtocol.HttpProtobuf` it knows for sure.

utpilla replied:
> The test really can't know for certain if the protocol was applied or just the default was used.

I think that's fine. This test should only be concerned about whether or not the exporter uses gRPC when the user calls `UseOtlpExporter` with the gRPC as the value for protocol parameter. We need not worry about whether it used the default value or rewrite the same value for this test.

oddparity commented:
Hi. Could it be this leads to a NotSupportedException, if (for reasons) my code calls the AddOtlpExporter method? The UseOtlpExporter method is called by this call tree:

new ElasticOpenTelemetryBuilder() at ElasticOpenTelemetryBuilder.cs:line 91
ServiceCollectionExtensions.AddElasticOpenTelemetry() at ServiceCollectionExtensions.cs:line 68
ServiceCollectionExtensions.AddElasticOpenTelemetry() at ServiceCollectionExtensions.cs:line 28
OpenTelemetryServicesExtensions.AddOpenTelemetry() at OpenTelemetryServicesExtensions.cs:line 28
CustomOpenTelemetry.AddCustomOpenTelemetryTracing() at .../CustomOpenTelemetry.cs:line 63
CustomOpenTelemetry.AddCustomOpenTelemetry()
Dependencies.AddInfrastructure()
Program.CreateWebApplication()
async Program.Main()
AsyncMethodBuilderCore.Start<Focus.Desktop.Program.<Main>d__1>()
AsyncTaskMethodBuilder<int>.Start<Focus.Desktop.Program.<Main>d__1>()
Program.Main()
Program.<Main>()

In line 63 of class CustomOpenTelemetry we call serviceCollection.AddOpenTelemetry().WithTracing(...). And of course the configuration delegate calls AddOtlpExporter() at the end. This looks more or less like the example code of your documentation. So UseOtlpExporter is not called by me. But with

		if (!options.DistroOptions.SkipOtlpExporter)
			openTelemetry.UseOtlpExporter();

it is called by your code. And this leads to:
System.NotSupportedException: Signal-specific AddOtlpExporter methods and the cross-cutting UseOtlpExporter method being invoked on the same IServiceCollection is not supported.

So what can I do?
